### PR TITLE
Announcements

### PR DESCRIPTION
**Pull request by prateekj117**
Add announcement entity, announcement repository, announcement management service, announcement management controller and announcement utils.

First step for #146

### CODE REVIEW

File: src/main/java/amu/zhcet/announcement/Announcement.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    @OneToOne(fetch = FetchType.LAZY)
+    private User sender;
+    @Size(max = 150)
+    private String title;
+    @NotBlank
+    @Size(max = 500)
+    private String message;
+    @NotNull
+    @Enumerated(EnumType.STRING)
+    private boolean scheduled;
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/announcement/Announcement.java]: It's a boolean, so no Enumerated annotation

File: src/main/java/amu/zhcet/announcement/Announcement.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    private LocalDateTime sentTime = LocalDateTime.now();
+    private boolean automated;
+
+    private transient int seenCount;
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/announcement/Announcement.java]: Not needed

File: src/main/java/amu/zhcet/announcement/Announcement.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    private transient boolean stopEmailPropagation;
+    private transient boolean stopFirebasePropagation;
+    private transient String icon = "https://zhcet-backend.firebaseapp.com/static/img/icon.png";
+    private transient Function<Announcement, LinkMessage> linkMessageConverter;
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/announcement/Announcement.java]: No transient field needed

File: src/main/java/amu/zhcet/announcement/management/AnnouncementManagementController.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        model.addAttribute("page_title", "Manage Announcements");
+        model.addAttribute("page_subtitle", "Announcement Manager");
+        model.addAttribute("page_description", "Manage and monitor sent announcements");
+
+        int currentPage = AnnouncementUtils.normalizePage(page);
+        Page<Announcement> announcementPage = announcementManagementService.getAnnouncements(currentPage);
+
+        AnnouncementUtils.prepareAnnouncements(model, announcementPage, currentPage);
+        List<Announcement> notifications = announcementPage.getContent();
+        model.addAttribute("notifications", notifications);
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/announcement/management/AnnouncementManagementController.java]: announcements

File: src/main/java/amu/zhcet/common/utils/AnnouncementUtils.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    public static void prepareAnnouncements(Model model, Page<?> page, int currentPage) {
+        int totalPages = page.getTotalPages();
+        int minPage = getSafePage(currentPage - 5, totalPages);
+        int maxPage = getSafePage(currentPage + 5, totalPages);
+
+        model.addAttribute("minPage", minPage);
+        model.addAttribute("maxPage", maxPage);
+        model.addAttribute("currentPage", getSafePage(currentPage, totalPages));
+    }
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/common/utils/AnnouncementUtils.java]: We'll reuse notification utils, so no need of recreating this class

File: src/main/java/amu/zhcet/announcement/Announcement.java (hunk 2 above)
Comment (prateekj117) [reply src/main/java/amu/zhcet/announcement/Announcement.java]: Localtime will be required I think.

Comment (iamareebjamal) [reply src/main/java/amu/zhcet/announcement/Announcement.java]: My comment was about seenCount

File: src/main/java/amu/zhcet/announcement/Announcement.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.function.Function;
+
+@Data
+@Entity
+@Audited
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode(callSuper = false)
+@ToString()
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/announcement/Announcement.java]: Not needed

File: src/main/resources/db/migration/V7.0.0__add_announcement_schema.sql [hunk 7]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE announcement
```
Comment (iamareebjamal) [inline src/main/resources/db/migration/V7.0.0__add_announcement_schema.sql]: Missing AUD details and table

File: src/main/java/amu/zhcet/announcement/AnnouncementRepository.java [hunk 8]
```diff
@@ -2,10 +2,11 @@
 
 import org.springframework.data.domain.Page;
 import org.springframework.data.domain.Pageable;
-import org.springframework.data.repository.PagingAndSortingRepository;
+import org.springframework.data.jpa.repository.JpaRepository;
 
-public interface AnnouncementRepository extends PagingAndSortingRepository<Announcement, Long> {
+public interface AnnouncementRepository extends JpaRepository<Announcement, Long> {
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/announcement/AnnouncementRepository.java]: Revert this

File: src/main/java/amu/zhcet/announcement/AnnouncementRepository.java [hunk 9]
```diff
@@ -2,10 +2,11 @@
...
 
     Page<Announcement> findBySenderUserId(String userId, Pageable Page);
+    Page<Announcement> findAll(String userId, Pageable pageable);
```
Comment (iamareebjamal) [inline src/main/java/amu/zhcet/announcement/AnnouncementRepository.java]: Add a new line above this

**Comment by iamareebjamal** [comment]
@prateekj117 Please squash your commits so I can review this easily

**Pull request merged**